swarm/src/handler.rs

```diff
@@ -379,11 +422,21 @@ pub enum ConnectionHandlerEvent<TConnectionUpgrade, TOutboundOpenInfo, TCustom,
     /// the connection, return [`KeepAlive::No`] in
     /// [`ConnectionHandler::connection_keep_alive`].
     Close(TErr),
+    /// We learned something about the protocols supported by the remote.
+    ReportRemoteProtocols(ProtocolSupport),
```

Comment:
We also need a `ReportLocalProtocols` event, right?

In the case of `libp2p-kad` when changing from client mode to server mode the `libp2p-kad` `ConnectionHandler` needs to emit a `ConnectionHandlerEvent::ReportLocalProtocols`.

Now one could argue that the change would be visible with the next in- or outbound stream, as `libp2p-swarm` collects the local supported protocols via `UpgradeInfo::protocol_info`. Though it is a bit tricky with Kademlia client mode. Say that the local node does not do any outgoing streams. When the local mode is in client mode, remote peers would not open a stream to it, thus also no inbound streams. Now when switching from client to server mode the local node will not get any incoming streams until it advertises support for it via libp2p-identify. Though that won't ever happen as `UpgradeInfo::protocol_info` is never called.

Does that make sense?

Reply:
What if we call `listen_protocol` and compute the diff on each step of the `Connection` event loop instead?

Something must trigger the switch in protocols so if we always check for listen protocols, we can do it transparently in the background.

The only risk here is that the user produces an endless loop if they keep changing their protocols as part of getting a changed protocols event but I think we can detect that.

Reply:
> What if we call `listen_protocol` and compute the diff on each step of the `Connection` event loop instead?

:+1: in my eyes that would significantly reduce complexity for the user.

Based on intuition it seems like an expensive operation to be potentially executed on each `Connection::poll` though, right? Even if we remove the in- and out-bound upgrade mechanism, it would still be a `collect` on an `Iterator` into a `HashSet` as well as a comparison to an existing `HashSet`. Am I missing something?

Reply:
If we do it as the last step of `poll`, it should only happen once everything else is idle, right? That shouldn't impact performance much.

Reply:
:+1: let's do it at the end of the `poll`.